File: parser/SimpleParser.java
/*=============================================================================
 * parser for pseudo-Boolean instances
 * 
 * Copyright (c) 2005-2007 Olivier ROUSSEL and Vasco MANQUINHO
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *=============================================================================
 */

// version 2.9.0

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.math.BigInteger;
import java.util.Vector;
import java.util.Iterator;
import java.lang.Integer;

class SimpleParser {
    /*
     * This protected section defines the callback that are used by the parser
     * to send data back to the program.
     * 
     * These functions must be redefined in a subclass.
     */

    /**
     * callback called when we get the number of variables and the expected
     * number of constraints
     * 
     * @param nbvar:
     *            the number of variables
     * @param nbconstr:
     *            the number of contraints
     */
    protected void metaData(int nbvar, int nbconstr) {
        System.out.println("[nbvar=" + nbvar + "]");
        System.out.println("[nbconstr=" + nbconstr + "]");
    }

    /**
     * callback called before we read the objective function
     */
    protected void beginObjective() {
        System.out.print("objective:  ");
    }

    /**
     * callback called after we've read the objective function
     */
    protected void endObjective() {
        System.out.println();
    }

    /**
     * callback called when we read a term of the objective function
     * 
     * @param coeff:
     *            the coefficient of the term
     * @param idVar: 
     *            the numerical identifier of the variable
     */
    protected void objectiveTerm(BigInteger coeff, int idVar) {
        System.out.print("[" + (coeff.signum() == 1 ? "+" : "") + coeff + " x"
                + idVar + "] ");
    }

   /**
     * callback called when we read a term of the objective function
     * which is a product of literals
     * 
     * @param coeff:
     *            the coefficient of the term
     * @param list: 
     *            list of literals which appear in the product
     */
    protected void objectiveProduct(BigInteger coeff, Vector list) {
        System.out.print("[" + (coeff.signum() == 1 ? "+" : "") + coeff + " ");

	Iterator i=list.iterator();

	while (i.hasNext())
	{
	    int id=((Integer)i.next()).intValue();
	    
	    if (id<0)
		System.out.print("~x"+(-id)+" ");
	    else
		System.out.print("x"+id+" ");
	}

        System.out.print("] ");
    }

    /**
     * callback called before we read a constraint
     */
    protected void beginConstraint() {
        System.out.print("constraint: ");
    }

    /**
     * callback called after we've read a constraint
     */
    protected void endConstraint() {
        System.out.println();
    }

    /**
     * callback called when we read a term of a constraint
     * 
     * @param coeff:
     *            the coefficient of the term
     * @param idVar:
     *            the numerical identifier of the variable
     */
    protected void constraintTerm(BigInteger coeff, int idVar) {
        System.out.print("[" + (coeff.signum() == 1 ? "+" : "") + coeff + " x"
                + idVar + "] ");
    }

    /**
     * callback called when we read a term of a constraint
     * which is a product of literals
     * 
     * @param coeff:
     *            the coefficient of the term
     * @param list: 
     *            list of literals which appear in the product
     */
    protected void constraintProduct(BigInteger coeff, Vector list) {
        System.out.print("[" + (coeff.signum() == 1 ? "+" : "") + coeff + " ");

	Iterator i=list.iterator();

	while (i.hasNext())
	{
	    int id=((Integer)i.next()).intValue();
	    
	    if (id<0)
		System.out.print("~x"+(-id)+" ");
	    else
		System.out.print("x"+id+" ");
	}

        System.out.print("] ");
    }

    /**
     * callback called when we read the relational operator of a constraint
     * 
     * @param relop:
     *            the relational oerator (>= or =)
     */
    protected void constraintRelOp(String relop) {
        System.out.print("[" + relop + "] ");
    }

    /**
     * callback called when we read the right term of a constraint (also known
     * as the degree)
     * 
     * @param val:
     *            the degree of the constraint
     */
    protected void constraintRightTerm(BigInteger val) {
        System.out.print("[" + (val.signum() == 1 ? "+" : "") + val + "]");
    }

    BufferedReader in; // the stream we're reading from

    char savedChar; // a character read from the file but not yet consumed

    boolean charAvailable = false; // true iff savedChar contains a character

    boolean eofReached = false; // true iff we've reached EOF
    
    int nbVars, nbConstr; // MetaData: #Variables and #Constraints in file.
    
    public class ParseException extends Exception {
        /**
         * Constructor for ParseException.
         * 
         * @param message
         *            the error message
         */
        public ParseException(String message) {
            super(message);
        }
    }

    /**
     * get the next character from the stream
     * 
     * @throws IOException
     */
    private char get() throws IOException {
        int c;

        if (charAvailable) {
            charAvailable = false;
            return savedChar;
        }

        c = in.read();
        if (c == -1)
            eofReached = true;

        return (char) c;
    }

    /**
     * put back a character into the stream (only one chr can be put back)
     */
    private void putback(char c) {
        savedChar = c;
        charAvailable = true;
    }

    /**
     * return true iff we've reached EOF
     */
    private boolean eof() {
        return eofReached;
    }

    /**
     * skip white spaces
     * 
     * @throws IOException
     */
    private void skipSpaces() throws IOException {
        char c;

        while (Character.isWhitespace(c = get()))
            ;

        putback(c);
    }

    /**
     * read a word from file
     * 
     * @return the word we read
     * @throws IOException
     */
    public String readWord() throws IOException {
        StringBuffer s = new StringBuffer();
        char c;

        skipSpaces();

        while (!Character.isWhitespace(c = get()) && !eof())
            s.append(c);

        return s.toString();
    }

    /**
     * read a integer from file
     * 
     * @param s:
     *            a StringBuffer to store the integer that was read
     * @throws IOException
     */
    public void readInteger(StringBuffer s) throws IOException {
        char c;

        skipSpaces();
        s.setLength(0);

        c = get();
        if (c == '-' || Character.isDigit(c))
            s.append(c);
        // note: BigInteger don't like a '+' before the number, we just skip it

        while (Character.isDigit(c = get()) && !eof())
            s.append(c);

        putback(c);
    }

    /**
     * read an identifier from stream and store it in s
     * 
     * @return the identifier we read or null
     * @throws IOException, ParseException
     */
    private boolean readIdentifier(Vector list) throws IOException, ParseException {
	StringBuffer s=new StringBuffer();
        char c;
	boolean negated=false;
	
        skipSpaces();

        // first char (must be a 'x')
        c = get();
        if (eof())
            return false;
	
	if (c == '~')
	{
	    negated=true;
	    c=get();
	}

        if (c != 'x') {
            putback(c);
            return false;
        }

        s.append(c);

        // next chars (must be digits)
        while (true) {
            c = get();
            if (eof())
                break;

            if (Character.isDigit(c))
                s.append(c);
            else {
                putback(c);
                break;
            }
        }
	
	//Small check on the coefficient ID to make sure everything is ok
	int varID = Integer.parseInt(s.substring(1));
	if (varID > nbVars) {
	    throw new ParseException("Variable identifier larger than #variables in metadata.");
	}
	
	if (negated)
	    varID=-varID;

	list.add(new Integer(varID));
	
        return true;
    }

    /**
     * read a relational operator from stream and store it in s
     * 
     * @return the relational operator we read or null
     * @throws IOException
     */
    private String readRelOp() throws IOException {
        char c;

        skipSpaces();

        c = get();
        if (eof())
            return null;

        if (c == '=')
            return "=";

        if (c == '>' && get() == '=')
            return ">=";

        return null;
    }

    /**
     * read the first comment line to get the number of variables and the number
     * of constraints in the file
     * 
     * calls metaData with the data that was read
     * 
     * @throws IOException
     * @throws ParseException
     */
    private void readMetaData() throws IOException, ParseException {
        char c;
        String s;

        // get the number of variables and constraints
        c = get();
        if (c != '*')
            throw new ParseException(
                    "First line of input file should be a comment");

        s = readWord();
        if (eof() || !s.equals("#variable="))
            throw new ParseException(
                    "First line should contain #variable= as first keyword");

        nbVars = Integer.parseInt(readWord());

        s = readWord();
        if (eof() || !s.equals("#constraint="))
            throw new ParseException(
                    "First line should contain #constraint= as second keyword");

        nbConstr = Integer.parseInt(readWord());

        // skip the rest of the line
        in.readLine();

        // callback to transmit the data
        metaData(nbVars, nbConstr);
    }

    /**
     * skip the comments at the beginning of the file
     * 
     * @throws IOException
     */
    private void skipComments() throws IOException {
        char c = ' ';

        // skip further comments

        while (!eof() && (c = get()) == '*') {
            in.readLine();
        }

        putback(c);
    }

    /**
     * read a term into coeff and var
     * 
     * @param coeff:
     *            the coefficient of the variable
     * @param list:
     *            the list of literals identifiers in the product
     * @throws IOException
     * @throws ParseException
     */
    private void readTerm(StringBuffer coeff, Vector list)
            throws IOException, ParseException {
        char c;

	list.clear();

        readInteger(coeff);

        skipSpaces();

        while(readIdentifier(list));

	if (list.size()==0)
            throw new ParseException("identifier expected");
    }

    /**
     * read the objective line (if any)
     * 
     * calls beginObjective, objectiveTerm and endObjective
     * 
     * @throws IOException
     * @throws ParseException
     */
    private void readObjective() throws IOException, ParseException {
        char c;
        Vector list = new Vector();
        StringBuffer coeff = new StringBuffer();

        // read objective line (if any)

        skipSpaces();
        c = get();
        if (c != 'm') {
            // no objective line
            putback(c);
            return;
        }

        if (get() == 'i' && get() == 'n' && get() == ':') {
            beginObjective(); // callback

            while (!eof()) {
                readTerm(coeff, list);
		if (list.size()==1)
		{
		    int val=((Integer)list.firstElement()).intValue();
		    if (val>0)
			objectiveTerm(new BigInteger(coeff.toString()), 
				      val); // callback
		    else
			objectiveProduct(new BigInteger(coeff.toString()), 
					 list); // callback
		}
		else
		    objectiveProduct(new BigInteger(coeff.toString()), 
				     list); // callback

                skipSpaces();
                c = get();
                if (c == ';')
                    break; // end of objective
                else if (c == '-' || c == '+' || Character.isDigit(c))
                    putback(c);
                else
                    throw new ParseException(
                            "unexpected character in objective function");
            }

            endObjective();
        } else
            throw new ParseException("input format error: 'min:' expected");
    }

    /**
     * read a constraint
     * 
     * calls beginConstraint, constraintTerm and endConstraint
     * 
     * @throws ParseException
     * @throws IOException
     */
    private void readConstraint() throws IOException, ParseException {
        Vector list = new Vector();
        StringBuffer coeff = new StringBuffer();
        char c;

        beginConstraint();

        while (!eof()) {
	    readTerm(coeff, list);
	    if (list.size()==1) {
		int val=((Integer)list.firstElement()).intValue();
		if (val>0)
		    constraintTerm(new BigInteger(coeff.toString()), 
				   val); // callback
		else
		    constraintProduct(new BigInteger(coeff.toString()), 
				      list); // callback
	    }
	    else
		constraintProduct(new BigInteger(coeff.toString()), 
				  list); // callback

            skipSpaces();
            c = get();
            if (c == '>' || c == '=') {
                // relational operator found
                putback(c);
                break;
            } else if (c == '-' || c == '+' || Character.isDigit(c))
                putback(c);
            else
                throw new ParseException("unexpected character in constraint");
        }

        if (eof())
            throw new ParseException("unexpected EOF before end of constraint");

        String relop;
        if ((relop = readRelOp()) != null)
            constraintRelOp(relop);
        else
            throw new ParseException(
                    "unexpected relational operator in constraint");

        readInteger(coeff);
        constraintRightTerm(new BigInteger(coeff.toString()));

        skipSpaces();
        c = get();
        if (eof() || c != ';')
            throw new ParseException("semicolon expected at end of constraint");

        endConstraint();
    }

    /**
     * constructor which only opens the file
     * 
     * @throws FileNotFoundException
     */
    public SimpleParser(String filename) throws FileNotFoundException {
        in = new BufferedReader(new InputStreamReader(new FileInputStream(
                filename)));
    }

    /**
     * parses the file and uses the callbacks to send to send the data back to
     * the program
     * 
     * @throws IOException
     * @throws ParseException
     */
    public void parse() throws IOException, ParseException {
	char c;

        readMetaData();

        skipComments();

        readObjective();

        // read constraints
	int nbConstraintsRead = 0;
        while (!eof()) {
            skipSpaces();
            if (eof())
                break;

	    c=get();
	    putback(c);
	    if(c=='*')
		skipComments();

            if (eof())
                break;

            readConstraint();
	    nbConstraintsRead++;
        }
	
	//Small check on the number of constraints
	if (nbConstraintsRead != nbConstr) {
	    throw new ParseException("Number of constraints read is different from metadata.");
	}
    }

    public static void main(String[] args) {
        try {
            if (args.length != 1)
                System.out.println("usage: SimpleParser <filename>");
            else {
                SimpleParser parser = new SimpleParser(args[0]);

                parser.parse();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}; // class SimpleParser

File: parser/SimpleParser.cc
/*=============================================================================
 * parser for pseudo-Boolean instances
 * 
 * Copyright (c) 2005-2007 Olivier ROUSSEL and Vasco MANQUINHO
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *=============================================================================
 */

// version 2.9.4

#include <iostream>
#include <fstream>
#include <iomanip>
#include <stdexcept>
#include <vector>
#include <cassert>
#include "SimpleParser.h"
#include "DefaultCallback.h"
#include "AbcCallback.h"
using namespace std;

/**
 * this class stores products of literals (as a tree) in order to
 * associate unique identifiers to these product (for linearization)
 */
template< typename Callback >
class ProductStore {
 private:
  // we represent each node of a n-ary tree by a vector<ProductNode>
  struct ProductNode {
    int lit; // ID of the literal
    int productId; // identifier associated to the product of the
    // literals found from the root up to this node
    vector< ProductNode >* next; // list of next literals in a product

    ProductNode( int l ) {
      lit = l;
      productId = 0;
      next = NULL;
    }

    // if we define a destructor to free <next>, we'll have to define
    // a copy constructor and use reference counting. It's not worth it.
  };

  vector< ProductNode > root; // root of the n-ary tree
  int nextSymbol; // next available variable

  /**
   * define an order on ProductNode based on the literal (used to
   * speed up look up)
   */
  class ProductNodeLessLit {
   public:
    bool operator()( const ProductNode& a, const ProductNode& b ) {
      return a.lit < b.lit;
    }
  };

 public:
  /**
   * give the first extra variable that can be used
   */
  void setFirstExtraSymbol( int id ) {
    nextSymbol = id;
  }

  /**
   * get the identifier associated to a product term (update the list
   * if necessary)
   */
  int getProductVariable( vector< int >& list ) {
    vector< ProductNode >* p = &root;
    typename vector< ProductNode >::iterator pos;

    // list must be sorted
    sort( list.begin(), list.end() );

    // is this a known product ?
    for ( int i = 0; i < list.size(); ++i ) {
      assert( p != NULL );

      // look for list[i] in *p
      pos = lower_bound( p->begin(), p->end(), list[i], ProductNodeLessLit() );
      if ( pos == p->end() || ( *pos ).lit != list[i] )
        pos = p->insert( pos, ProductNode( list[i] ) ); // insert at the right place

      if ( i != list.size() - 1 && ( *pos ).next == NULL )
        ( *pos ).next = new vector< ProductNode >;

      p = ( *pos ).next;
    }

    if ( ( *pos ).productId == 0 )
      ( *pos ).productId = nextSymbol++;

    return ( *pos ).productId;
  }

  /**
   * add the constraints which define all product terms
   *
   */
  void defineProductVariable( Callback& cb ) {
    vector< int > list;

    defineProductVariableRec( cb, root, list );
  }

  /**
   * free all allocated product data
   *
   */
  void freeProductVariable() {
    freeProductVariableRec( root );
  }

 private:
  /**
   * add the constraints which define all product terms
   *
   */
  void defineProductVariableRec( Callback& cb,
                                 vector< ProductNode >& nodes, vector< int >& list ) {
    for ( int i = 0; i < nodes.size(); ++i ) {
      list.push_back( nodes[i].lit );
      if ( nodes[i].productId )
        cb.linearizeProduct( nodes[i].productId, list );

      if ( nodes[i].next )
        defineProductVariableRec( cb, *nodes[i].next, list );

      list.pop_back();
    }
  }

  /**
   * free all allocated product data
   *
   */
  void freeProductVariableRec( vector< ProductNode >& nodes ) {
    for ( int i = 0; i < nodes.size(); ++i ) {
      if ( nodes[i].next ) {
        freeProductVariableRec( *nodes[i].next );
        delete nodes[i].next;
      }
    }

    nodes.clear();
  }

};

template< typename Callback >
class SimpleParser {
 public:
  Callback cb;

 private:
  ifstream in; // the stream we're reading from
  int nbVars, nbConstr; // MetaData: #Variables and #Constraints in file.

  int nbProduct, sizeProduct; // MetaData for non linear format
  ProductStore< Callback > store;
  bool autoLinearize; // should the parser linearize constraints ?

  /**
   * get the next character from the stream
   */
  char get() {
    return in.get();
  }

  /**
   * put back a character into the stream (only one chr can be put back)
   */
  void putback( char c ) {
    in.putback( c );
  }

  /**
   * return true iff we've reached EOF
   */
  bool eof() {
    return !in.good();
  }

  /**
   * skip white spaces
   */
  void skipSpaces() {
    char c;

    while ( isspace( c = get() ) );

    putback( c );
  }

  /**
   * read an identifier from stream and append it to the list "list"
   * @param list: the current list of identifiers that were read
   * @return true iff an identifier was correctly read
   */
  bool readIdentifier( vector< int >& list ) {
    char c;
    bool negated = false;

    skipSpaces();

    // first char (must be 'x')
    c = get();
    if ( eof() )
      return false;

    if ( c == '~' ) {
      negated = true;
      c = get();
    }

    if ( c != 'x' ) {
      putback( c );
      return false;
    }

    int varID = 0;

    // next chars (must be digits)
    while ( true ) {
      c = get();
      if ( eof() )
        break;

      if ( isdigit( c ) )
        varID = varID * 10 + c - '0';
      else {
        putback( c );
        break;
      }
    }

    //Small check on the coefficient ID to make sure everything is ok
    if ( varID > nbVars )
      throw runtime_error( "variable identifier larger than "
                           "#variables in metadata." );

    if ( negated )
      varID = -varID;

    list.push_back( varID );

    return true;
  }

  /**
   * read a relational operator from stream and store it in s
   * @param s: the variable to hold the relational operator we read
   * @return true iff a relational operator was correctly read
   */
  bool readRelOp( string& s ) {
    char c;

    skipSpaces();

    c = get();
    if ( eof() )
      return false;

    if ( c == '=' ) {
      s = "=";
      return true;
    }

    if ( c == '>' && get() == '=' ) {
      s = ">=";
      return true;
    }

    return false;
  }

  /**
   * read the first comment line to get the number of variables and
   * the number of constraints in the file
   *
   * calls metaData with the data that was read
   */
  void readMetaData() {
    char c;
    string s;

    // get the number of variables and constraints
    c = get();
    if ( c != '*' )
      throw runtime_error( "First line of input file should be a comment" );

    in >> s;
    if ( eof() || s != "#variable=" )
      throw runtime_error( "First line should contain #variable= as first keyword" );

    in >> nbVars;
    store.setFirstExtraSymbol( nbVars + 1 );

    in >> s;
    if ( eof() || s != "#constraint=" )
      throw runtime_error( "First line should contain #constraint= as second keyword" );

    in >> nbConstr;

    skipSpaces();

    c = get();
    putback( c );

    if ( c == '#' ) {
      // assume non linear format
      in >> s;
      if ( eof() || s != "#product=" )
        throw runtime_error( "First line should contain #product= as third keyword" );

      in >> nbProduct;

      in >> s;
      if ( eof() || s != "sizeproduct=" )
        throw runtime_error( "First line should contain sizeproduct= as fourth keyword" );

      in >> sizeProduct;
    }

    // skip the rest of the line
    getline( in, s );

    // callback to transmit the data
    if ( nbProduct && autoLinearize ) {
#ifdef ONLYCLAUSES
      cb.metaData(nbVars+nbProduct,nbConstr+nbProduct+sizeProduct);
#else
      cb.metaData( nbVars + nbProduct, nbConstr + 2 * nbProduct );
#endif
    } else
      cb.metaData( nbVars, nbConstr );
  }

  /**
   * skip the comments at the beginning of the file
   */
  void skipComments() {
    string s;
    char c;

    // skip further comments

    while ( !eof() && ( c = get() ) == '*' ) {
      getline( in, s );
    }

    putback( c );
  }

  /**
   * read a term into coeff and list
   * @param coeff: the coefficient of the variable
   * @param list: the list of literals identifiers in the product
   */
  void readTerm( IntegerType& coeff, vector< int >& list ) {
    char c;

    list.clear();

    in >> coeff;

    skipSpaces();

    while ( readIdentifier( list ) );

    if ( list.size() == 0 )
      throw runtime_error( "identifier expected" );
  }

  /**
   * read the objective line (if any)
   *
   * calls beginObjective, objectiveTerm and endObjective
   */
  void readObjective() {
    char c;
    string s;

    IntegerType coeff;
    vector< int > list;

    // read objective line (if any)

    skipSpaces();
    c = get();
    if ( c != 'm' ) {
      // no objective line
      putback( c );
      return;
    }

    if ( get() == 'i' && get() == 'n' && get() == ':' ) {
      cb.beginObjective(); // callback

      while ( !eof() ) {
        readTerm( coeff, list );
        if ( list.size() == 1 && list[0] > 0 )
          cb.objectiveTerm( coeff, list[0] );
        else
          handleProduct( true, coeff, list );

        skipSpaces();
        c = get();
        if ( c == ';' )
          break; // end of objective
        else if ( c == '-' || c == '+' || isdigit( c ) )
          putback( c );
        else
          throw runtime_error( "unexpected character in objective function" );
      }

      cb.endObjective();
    } else
      throw runtime_error( "input format error: 'min:' expected" );
  }

  /**
   * read a constraint
   *
   * calls beginConstraint, constraintTerm and endConstraint
   */
  void readConstraint() {
    string s;
    char c;

    IntegerType coeff;
    vector< int > list;

    cb.beginConstraint();

    while ( !eof() ) {
      readTerm( coeff, list );
      if ( list.size() == 1 && list[0] > 0 )
        cb.constraintTerm( coeff, list[0] );
      else
        handleProduct( false, coeff, list );

      skipSpaces();
      c = get();
      if ( c == '>' || c == '=' ) {
        // relational operator found
        putback( c );
        break;
      } else if ( c == '-' || c == '+' || isdigit( c ) )
        putback( c );
      else
        throw runtime_error( "unexpected character in constraint" );
    }

    if ( eof() )
      throw runtime_error( "unexpected EOF before end of constraint" );

    if ( readRelOp( s ) )
      cb.constraintRelOp( s );
    else
      throw runtime_error( "unexpected relational operator in constraint" );

    in >> coeff;
    cb.constraintRightTerm( coeff );

    skipSpaces();
    c = get();
    if ( eof() || c != ';' )
      throw runtime_error( "semicolon expected at end of constraint" );

    cb.endConstraint();
  }

  /**
   * passes a product term to the solver (first linearizes the product
   * if this is wanted)
   */
  void handleProduct( bool inObjective, IntegerType coeff, vector< int >& list ) {
    if ( autoLinearize ) {
      // get symbol corresponding to this product
      int var = store.getProductVariable( list );

      if ( inObjective )
        cb.objectiveTerm( coeff, var );
      else
        cb.constraintTerm( coeff, var );
    } else {
      if ( inObjective )
        cb.objectiveProduct( coeff, list );
      else
        cb.constraintProduct( coeff, list );
    }
  }

 public:
  /**
   * constructor which only opens the file
   */
  SimpleParser( char* filename ) {
    in.open( filename, ios_base::in );

    if ( !in.good() )
      throw runtime_error( "error opening input file" );

    autoLinearize = false;

    nbVars = 0;
    nbConstr = 0;
    nbProduct = 0;
    sizeProduct = 0;
  }

  ~SimpleParser() {
    store.freeProductVariable();
  }

  /**
   * ask the parser to linearize the constraints using a simple
   * default method
   */
  void setAutoLinearize( bool lin = true ) {
    autoLinearize = lin;
  }

  /**
   * parses the file and uses the callbacks to send the data
   * back to the program
   */
  void parse() {
    char c;

    readMetaData();

    skipComments();

    readObjective();

    // read constraints
    int nbConstraintsRead = 0;
    while ( !eof() ) {
      skipSpaces();
      if ( eof() )
        break;

      putback( c = get() );
      if ( c == '*' )
        skipComments();

      if ( eof() )
        break;

      readConstraint();
      nbConstraintsRead++;
    }

    //Small check on the number of constraints
    if ( nbConstraintsRead != nbConstr )
      throw runtime_error( "number of constraints read is different from metadata." );

    if ( autoLinearize ) {
      store.defineProductVariable( cb );
    }
  }
}; // class SimpleParser


int main( int argc, char* argv[] ) {
  try {
    if ( argc != 2 )
      cout << "usage: SimpleParser <filename>" << endl;
    else {
      SimpleParser< AbcCallback > parser( argv[1] );

      parser.setAutoLinearize( true );
      parser.parse();
    }
  }
  catch ( exception& e ) {
    cout.flush();
    cerr << "ERROR: " << e.what() << endl;
  }

  return 0;
}
